diff -Naur rpmtools-6.1/genhdlist2 rpmtools-6.1.tpg/genhdlist2
--- rpmtools-6.1/genhdlist2	2011-05-31 11:20:48.000000000 +0000
+++ rpmtools-6.1.tpg/genhdlist2	2013-01-31 19:31:16.618043346 +0000
@@ -17,7 +17,7 @@
     my %options = (
 	synthesis_filter => '.cz:gzip -9',
 	xml_info => 'auto',
-	xml_info_filter => '.lzma:lzma -5',
+	xml_info_filter => '.lzma:lzma -5 --text',
 	versioned => 'auto'
     );
 
